magefile.go

```diff
@@ -0,0 +1,97 @@
+//go:build mage
+// +build mage
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+
+	"github.com/mt-sre/devkube/magedeps"
+)
+
+// Directories
+var (
+	// Working directory of the project.
+	workDir string
+	// Cache directory for temporary build files.
+	cacheDir string
+	// Dependency directory.
+	depsDir magedeps.DependencyDirectory
+)
+
+func init() {
+	var err error
+	// Directories
+	workDir, err = os.Getwd()
+	if err != nil {
+		panic(fmt.Errorf("getting work dir: %w", err))
+	}
+
+	depsDir = magedeps.DependencyDirectory(path.Join(workDir, ".deps"))
+	cacheDir = path.Join(workDir, ".cache")
+
+	// Path
+	os.Setenv("PATH", depsDir.Bin()+":"+os.Getenv("PATH"))
+}
+
+// Runs go mod tidy in all go modules of the repository.
+func Tidy() error {
+	if err := sh.Run("go", "mod", "tidy"); err != nil {
+		return fmt.Errorf("tidy main module: %w", err)
+	}
+	return nil
+}
+
+// Testing and Linting
+// -------------------
+
+type Test mg.Namespace
+
+// Runs unittests.
+func (Test) Unit() error {
+	return sh.RunWithV(map[string]string{
+		"CGO_ENABLED": "1",
+	}, "go", "test", "-v", "-race", "./dev/...", "./cmd/...", "./magedeps/...")
+}
+
+// Lints the source code.
+func (Test) Lint() error {
+	mg.Deps(Dependency.GolangciLint)
+
+	for _, cmd := range [][]string{
+		{"go", "fmt", "./..."},
+		{"golangci-lint", "run", "./...", "--deadline=15m"},
```

Comment:
`ctx.Deadline` returns a `time.Time` not a `time.Duration`, so it's not as trivial as your code snipped.

I would really like for the mage sh package to use context itself, so the exec process would get a SIGKill to cleanup:
https://pkg.go.dev/os/exec#CommandContext

Reply:
That's why I used `time.Until` which as you're pointing out is part of the reason it doesn't work super great. Mage just exits instead of waiting for child processes to end which is always an issue when those subprocesses timeout exactly when mage intends to exit..